lib/lumberjack/server.rb

```diff
@@ -36,49 +39,91 @@ def initialize(options={})
         end
       end
 
-      tcp_server = TCPServer.new(@options[:address], @options[:port])
+      @server = TCPServer.new(@options[:address], @options[:port])
+
+      @close = Concurrent::AtomicBoolean.new
 
       # Query the port in case the port number is '0'
       # TCPServer#addr == [ address_family, port, address, address ]
-      @port = tcp_server.addr[1]
+      @port = @server.addr[1]
 
-      if !@options[:ssl]
-        @server = tcp_server
-      else
+      if @options[:ssl]
         # load SSL certificate
-        ssl = OpenSSL::SSL::SSLContext.new
-        ssl.cert = OpenSSL::X509::Certificate.new(File.read(@options[:ssl_certificate]))
-        ssl.key = OpenSSL::PKey::RSA.new(File.read(@options[:ssl_key]),
+        @ssl = OpenSSL::SSL::SSLContext.new
+        @ssl.cert = OpenSSL::X509::Certificate.new(File.read(@options[:ssl_certificate]))
+        @ssl.key = OpenSSL::PKey::RSA.new(File.read(@options[:ssl_key]),
           @options[:ssl_key_passphrase])
-
-        @server = OpenSSL::SSL::SSLServer.new(tcp_server, ssl)
       end
-
     end # def initialize
 
     def run(&block)
-      while true
+      while !closed?
         connection = accept
-        Thread.new(connection) do |connection|
-          connection.run(&block)
+
+        # Some exception may occur in the accept loop
+        # we will try again in the next iteration
+        # unless the server is closing
+        if connection
```

Comment:
minor subjective style comment: I'd rather have `next unless connection`, then have the block be indented one level shallower.